flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/FlinkKafkaProducer.java

```diff
@@ -226,13 +228,19 @@ public int getTransactionCoordinatorId() {
 
 	private void flushNewPartitions() {
 		LOG.info("Flushing new partitions");
+		enqueueNewPartitions().await();
+	}
+
+	private TransactionalRequestResult enqueueNewPartitions() {
 		Object transactionManager = getValue(kafkaProducer, "transactionManager");
-		Object txnRequestHandler = invoke(transactionManager, "addPartitionsToTransactionHandler");
-		invoke(transactionManager, "enqueueRequest", new Class[]{txnRequestHandler.getClass().getSuperclass()}, new Object[]{txnRequestHandler});
-		TransactionalRequestResult result = (TransactionalRequestResult) getValue(txnRequestHandler, txnRequestHandler.getClass().getSuperclass(), "result");
-		Object sender = getValue(kafkaProducer, "sender");
-		invoke(sender, "wakeup");
-		result.await();
+		synchronized (transactionManager) {
+			Object txnRequestHandler = invoke(transactionManager, "addPartitionsToTransactionHandler");
+			invoke(transactionManager, "enqueueRequest", new Class[]{txnRequestHandler.getClass().getSuperclass()}, new Object[]{txnRequestHandler});
+			TransactionalRequestResult result = (TransactionalRequestResult) getValue(txnRequestHandler, txnRequestHandler.getClass().getSuperclass(), "result");
+			Object sender = getValue(kafkaProducer, "sender");
+			invoke(sender, "wakeup");
```

Comment:
`sender.wakeup` is outside of the lock in the original code. Do you think it makes a difference?

Reply:
Good catch. It shouldn't make a difference, but I will change it to better match original code.